### PR TITLE
Make Gizmos compatible with FixedUpdate

### PR DESCRIPTION
**Pull request by SpecificProtagonist**
Superseded by #10973

# Objective

Gizmos are immediate mode and thus last only for one frame. When drawing them during `FixedUpdate`, multiple frames may be drawn before the next fixed update, causing them to flicker.

## Solution

Record fixed update gizmos separately and keep them around till the next update. This also works if there are multiple fixed updates in a single frame.

Also expose `GizmoBuffer` for any custom scheduling not covered by this.

## Alternatives

#10973 achieves the same by adding a generic parameter with a default value to Gizmos. While *technically* less flexible for extension as users can only store gizmo buffers based on their type instead of as values, I think that's sufficient and is a solution I'd be happy with.

---

### Added
- `GizmoBuffer` for recording gizmo draw calls

### Fixed
- fixed gizmos flickering is drawn in `FixedUpdate`

### CODE REVIEW

**Comment by aevyrie** [comment]
Wouldn't it make more sense to have a retained API for gizmos, e.g. `Handle<Polyline>`, for cases like this?

File: crates/bevy_gizmos/src/lib.rs [hunk 1]
```diff
@@ -92,6 +95,11 @@ impl Plugin for GizmoPlugin {
                 ),
             );
 
+        // Ensure gizmos from prefious fixed update are cleaned up if no other system
```
Comment (Aceeri) [inline crates/bevy_gizmos/src/lib.rs:98]:
```suggestion
        // Ensure gizmos from previous fixed update are cleaned up if no other system
```

**Comment by Aceeri** [comment]
> Wouldn't it make more sense to have a retained API for gizmos, e.g. `Handle<Polyline>`, for cases like this?

I would say no mostly because I use these things as quick debug tools. Retained mode would add a lot of boilerplate here.

**Comment by rlidwka** [comment]
You are only fixing issue for fixed update, it wouldn't work for custom schedules, `run_if` conditions, and so on.

Perhaps gizmos should be retained {until next system run OR until 1 second is passed}? This solution would make gizmos work in all systems. But I'm not sure about performance implications of this approach.

**Review by mnmaita — COMMENTED** [review]
Leaving minor suggestions and typos on some doc comments.

**Comment by alice-i-cecile** [comment]
> Wouldn't it make more sense to have a retained API for gizmos, e.g. `Handle<Polyline>`, for cases like this?

This is sort of the direction that I lean towards. I don't think you should be drawing things to the screen during `FixedUpdate` at all, especially with an immediate mode API.

**Comment by aevyrie** [comment]
This is adding retained mode behavior to an immediate mode API. What if someone uses this expecting the gizmo to draw only on the frames the fixedupdate is running? That's how I would expect this to work, and this change would prevent that. Like, lets say I want to find the exact frame my updates are happening, visually, I would expect the gizmo would only display for the rendered frame the fixed update system is running. I could then record my screen to visually debug precisely when updates are happening.

You can make gizmos work the way you want without any changes:
```rs
// scheduled on a fixed update
fn update_pathfinding(mut pathfinding: ResMut<Pathfinding>) {}

// normally scheduled system
fn debug_pathfinding(mut gizmos: Gizmos, pathfinding: Res<Pathfinding>) {}
```

**Comment by Aceeri** [comment]
> This is adding retained mode behavior to an immediate mode API. What if someone uses this expecting the gizmo to draw only on the frames the fixedupdate is running? That's how I would expect this to work, and this change would prevent that. Like, lets say I want to find the exact frame my updates are happening, visually, I would expect the gizmo would only display for the rendered frame the fixed update system is running.

The problem I have with this is that now you have boilerplate and friction to just adding debugging utility. I can't just plop a line of code into my gameplay systems at any point and debug it similar to a `info!(...)`. The way I see it is `FixedUpdate` is where basically all gameplay systems *should* be, this makes it significantly easier to debug/have deterministic state/network/etc.

> I could then record my screen to visually debug precisely when updates are happening.

I'm unsure of the point of that... Even in networking situations I am not using visual feedback.

**Comment by alice-i-cecile** [comment]
Ultimately, drawing things, especially in immediate mode, in FixedUpdate is fundamentally questionable. I think that your idea of "just use it like dbg" is good (and that "gameplay systems should be in FIxedUpdate by default"), but there are some real challenges here.

FixedUpdate's behavior is *uniquely* strange, and with good reason. It may run once, zero or many times for a single frame. Once works fine out of the box. Your solution handles zero times, at the cost of substantial complexity. But what about multiple times? Your solution (and any non-retained solution) will then render *multiple* copies of the gizmos on the screen simultaneously, if only for a second.

Ultimately, I think that the best solution here is to say "sorry, you can't use immediate mode drawing APIs within FixedUpdate", and then clearly document why and the workaround (split it out into another system that runs in Update).

**Comment by Aceeri** [comment]
> Ultimately, drawing things, especially in immediate mode, in FixedUpdate is fundamentally questionable. I think that your idea of "just use it like dbg" is good (and that "gameplay systems should be in FIxedUpdate by default"), but there are some real challenges here.
> 
> FixedUpdate's behavior is _uniquely_ strange, and with good reason. It may run once, zero or many times for a single frame. Once works fine out of the box. Your solution handles zero times, at the cost of substantial complexity. But what about multiple times? Your solution (and any non-retained solution) will then render _multiple_ copies of the gizmos on the screen simultaneously, if only for a second.
> 
> Ultimately, I think that the best solution here is to say "sorry, you can't use immediate mode drawing APIs within FixedUpdate", and then clearly document why and the workaround (split it out into another system that runs in Update).

What if we just replicate the Time<Fixed> stuff here? I don't see why the workaround can't just be the solution here.

**Pull request closed without merging**